src/mug.gleam

```diff
@@ -354,67 +543,167 @@ pub type IpVersionPreference {
+/// sent as-is to the peer. If chain certificates are not provided, certificates from the configured trusted CA certificates
+/// will be used to construct the chain.
+///
+/// If verification is disabled, this function does nothing.
+///
+pub fn certificates_keys(
+  options: ConnectionOptions,
+  certificates_keys certificates_keys: List(CertificatesKeys),
+) -> ConnectionOptions {
+  ConnectionOptions(..options, tls_opts: case options.tls_opts {
+    UseTls(TlsOptions(verification_method)) ->
+      UseTls(
+        TlsOptions(verification_method: case verification_method {
+          Certificates(system, cacerts, _) ->
+            Certificates(system, cacerts, certificates_keys)
+          _ -> verification_method
+        }),
+      )
+    NoTls -> options.tls_opts
+  })
+}
+
 pub fn connect(options: ConnectionOptions) -> Result(Socket, ConnectError) {
   let host = charlist.from_string(options.host)
-  let connect = fn(inet) {
-    let gen_options = [
-      inet,
-      // When data is received on the socket queue it in the TCP stack rather than
-      // sending it as an Erlang message to the socket owner's inbox.
-      Active(passive()),
-      // We want the data from the socket as bit arrays please, not lists.
-      Mode(Binary),
-    ]
-    gen_tcp_connect(host, options.port, gen_options, options.timeout)
+  let connect = fn(use_inet6: Bool) {
+    let tcp_opts = get_tcp_options(use_inet6)
+    case options.tls_opts {
+      UseTls(TlsOptions(vm)) -> {
+        use opts <- result.try(get_tls_options(vm))
+        ssl_connect(host, options.port, tcp_opts, opts, options.timeout)
+        |> result.map(SslSocket)
+      }
+      NoTls -> {
+        gen_tcp_connect(host, options.port, tcp_opts, options.timeout)
+        |> result.map(TcpSocket)
+      }
+    }
   }
   case options.ip_version_preference {
-    Ipv4Only -> connect(Inet) |> result.map_error(ConnectFailedIpv4)
-    Ipv6Only -> connect(Inet6) |> result.map_error(ConnectFailedIpv6)
+    Ipv4Only -> connect(False) |> result.map_error(ConnectFailedIpv4)
+    Ipv6Only -> connect(True) |> result.map_error(ConnectFailedIpv6)
 
     Ipv4Preferred ->
-      case connect(Inet) {
+      case connect(False) {
         Ok(conn) -> Ok(conn)
         Error(ipv4) ->
-          case connect(Inet6) {
+          case connect(True) {
             Ok(conn) -> Ok(conn)
             Error(ipv6) -> Error(ConnectFailedBoth(ipv4:, ipv6:))
           }
       }
 
     Ipv6Preferred ->
-      case connect(Inet6) {
+      case connect(True) {
         Ok(conn) -> Ok(conn)
         Error(ipv6) ->
-          case connect(Inet) {
+          case connect(False) {
             Ok(conn) -> Ok(conn)
             Error(ipv4) -> Error(ConnectFailedBoth(ipv4:, ipv6:))
           }
       }
   }
 }
 
-type ModeValue {
-  Binary
+/// Returns the default gen_tcp options for mug
+fn get_tcp_options(use_inet6: Bool) -> List(GenTcpOption) {
```

Comment:
Go back to the previous design of passing in the inet variant please, do not use a bool